src/Microsoft.OpenApi.OData.Reader/Common/Utils.cs

```diff
@@ -270,5 +271,132 @@ internal static void AddCustomAttributesToExtensions(this IDictionary<string, IO
 
             return false;
         }
+
+
+        /// <summary>
+        /// Gets the entity type of the target <paramref name="segment"/>.
+        /// </summary>
+        /// <param name="segment">The target <see cref="ODataSegment"/>.</param>
+        /// <returns>The entity type of the target <paramref name="segment"/>.</returns>
+        internal static IEdmEntityType EntityTypeFromPathSegment(this ODataSegment segment)
+        {
+            CheckArgumentNull(segment, nameof(segment));
+
+            if (segment is ODataNavigationPropertySegment navPropSegment)
+            {
+                return navPropSegment.EntityType;
+            }
+            else if (segment is ODataNavigationSourceSegment navSourceSegment)
+            {
+                if (navSourceSegment.NavigationSource is IEdmEntitySet entitySet)
+                {
+                    return entitySet.EntityType();
+                }                    
+                else if (navSourceSegment.NavigationSource is IEdmSingleton singleton)
+                {
+                    return singleton.EntityType();
+                }                    
+            }
+            else if (segment is ODataKeySegment keySegment)
+            {
+                return keySegment.EntityType;
+            }
+            else if (segment is ODataOperationSegment)
+            {
+                return segment.EntityTypeFromOperationSegment();
+            }
```

Comment:
A switch statement would be nice to use here.
```suggestion
            switch (segment)
            {
                case ODataNavigationPropertySegment navPropSegment:
                    return navPropSegment.EntityType;
                case ODataNavigationSourceSegment navSourceSegment when navSourceSegment.NavigationSource is IEdmEntitySet entitySet:
                    return entitySet.EntityType();
                case ODataNavigationSourceSegment navSourceSegment when navSourceSegment.NavigationSource is IEdmSingleton singleton:
                    return singleton.EntityType();
                case ODataKeySegment keySegment:
                    return keySegment.EntityType;
                case ODataOperationSegment:
                    return segment.EntityTypeFromOperationSegment();
                default:
                    return null;
            }
```